jctools-experimental/src/main/java/org/jctools/channels/proxy/ProxyChannelFactory.java

```diff
@@ -1,9 +1,424 @@
+
+            // # W_OFFSET_DELTA = 4
+            // #FOREACH param in method
+            int wOffsetDelta = 4;
+            for (Class<?> parameterType : method.getParameterTypes()) {
+                // #PUSH: UnsafeAccess.UNSAFE.get[param.type](rOffset + #W_OFFSET_DELTA);
+                // #W_OFFSET_DELTA += if param.type in {long, double} 8 else 4;
+                getUnsafe(methodVisitor, parameterType, 4, wOffsetDelta);
+                wOffsetDelta += memorySize(parameterType);
+            }
+            // #END
+
+            // this.readRelease(rOffset);
+            readRelease(methodVisitor, 4);
+
+            // method.invoke(impl, <args>);
+            methodVisitor.visitMethodInsn(Opcodes.INVOKEINTERFACE,
+                    Type.getInternalName(iFace),
+                    method.getName(),
+                    Type.getMethodDescriptor(method),
+                    true);
+
+            // break;
+            methodVisitor.visitJumpInsn(Opcodes.GOTO, endOfSwitch);
+        }
+
+        // label <endOfSwitch>;
+        methodVisitor.visitLabel(endOfSwitch);
+
+        // i++;
+        methodVisitor.visitIincInsn(3, 1);
+
+        // goto <loopStart>;
+        methodVisitor.visitJumpInsn(Opcodes.GOTO, loopStart);
+        methodVisitor.visitLabel(loopEnd);
+
+        // return i;
+        methodVisitor.visitVarInsn(Opcodes.ILOAD, 3);
+        methodVisitor.visitInsn(Opcodes.IRETURN);
+
+        // size requirement is computed by ASM; complete method.
+        methodVisitor.visitMaxs(-1, -1);
+        methodVisitor.visitEnd();
+
+        // Add generic bridge method for erasure public int process (Object impl, int limit)
+        methodVisitor = classVisitor.visitMethod(Opcodes.ACC_BRIDGE | Opcodes.ACC_SYNTHETIC | Opcodes.ACC_PUBLIC,
+                "process",
+                "(" + Type.getDescriptor(Object.class) + "I)I",
+                null,
+                null);
+        methodVisitor.visitCode();
+
+        methodVisitor.visitVarInsn(Opcodes.ALOAD, 0);
+        methodVisitor.visitVarInsn(Opcodes.ALOAD, 1);
+        methodVisitor.visitTypeInsn(Opcodes.CHECKCAST, Type.getInternalName(iFace));
+        methodVisitor.visitVarInsn(Opcodes.ILOAD, 2);
+
+        methodVisitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL,
+                generatedName,
+                "process",
+                "(" + Type.getDescriptor(iFace) + "I)I",
+                false);
+
+        methodVisitor.visitInsn(Opcodes.IRETURN);
+
+        methodVisitor.visitMaxs(-1, -1);
+        methodVisitor.visitEnd();
+    }
+
+    private static void implementConstructor(ClassVisitor classVisitor) {
+        MethodVisitor methodVisitor = classVisitor.visitMethod(Opcodes.ACC_PUBLIC, "<init>", "(I)V", null, null);
+        methodVisitor.visitCode();
+
+        methodVisitor.visitVarInsn(Opcodes.ALOAD, 0);
+        methodVisitor.visitVarInsn(Opcodes.ILOAD, 1);
+        methodVisitor.visitIntInsn(Opcodes.BIPUSH, 13);
```

Comment:
Why is the messageSize hardcoded to 13? I see that the Pow2.align in the off heap ring buffer classes adjust this to 20.